[SW-2586] Expose Model Metrics as Objects on H2OMOJOModel in Scala API

mn-mikke opened the pull request:
Generated classes:
[H2OAnomalyMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144838/H2OAnomalyMetrics.scala.txt)
[H2OAutoEncoderMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144839/H2OAutoEncoderMetrics.scala.txt)
[H2OBinomialGLMMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144840/H2OBinomialGLMMetrics.scala.txt)
[H2OBinomialMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144841/H2OBinomialMetrics.scala.txt)
[H2OClusteringMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144842/H2OClusteringMetrics.scala.txt)
[H2OCommonMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144843/H2OCommonMetrics.scala.txt)
[H2OGLRMMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144844/H2OGLRMMetrics.scala.txt)
[H2OMultinomialGLMMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144845/H2OMultinomialGLMMetrics.scala.txt)
[H2OMultinomialMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144846/H2OMultinomialMetrics.scala.txt)
[H2OOrdinalGLMMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144847/H2OOrdinalGLMMetrics.scala.txt)
[H2OOrdinalMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144848/H2OOrdinalMetrics.scala.txt)
[H2OPCAMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144849/H2OPCAMetrics.scala.txt)
[H2ORegressionCoxPHMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144850/H2ORegressionCoxPHMetrics.scala.txt)
[H2ORegressionGLMMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144851/H2ORegressionGLMMetrics.scala.txt)
[H2ORegressionMetrics.scala.txt](https://github.com/h2oai/sparkling-water/files/7144852/H2ORegressionMetrics.scala.txt)

michalkurka commented on api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredMetricFields.scala:
```diff
@@ -0,0 +1,32 @@
+      "model_checksum",
+      "frame",
+      "frame_checksum",
+      "description",
```
I would keep this field, could be useful, eg. for CV metrics it explains that it is "N-fold cross-validation on training data (Metrics computed for combined holdout predictions)"

mn-mikke replied:
Ok, will remove `description` from the list of ignored fields. Thanks!

michalkurka commented on scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:
```diff
@@ -89,10 +99,16 @@ abstract class H2OMOJOModel
+
   def getCrossValidationMetrics(): Map[String, Double] = $(crossValidationMetrics)
 
+  def getCrossValidationMetricsObject(): H2OMetrics = $(crossValidationMetricsObject)
```
I think this would deserve documentation, explanation these are the combined CV holdout metrics

mn-mikke replied:
Ok, I will improve that.

michalkurka commented on scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:
```diff
@@ -89,10 +99,16 @@ abstract class H2OMOJOModel
   def getCrossValidationMetrics(): Map[String, Double] = $(crossValidationMetrics)
 
+  def getCrossValidationMetricsObject(): H2OMetrics = $(crossValidationMetricsObject)
+
```
Unrelated to this PR - How do we get the Metrics CV summary? is that exposed?

mn-mikke replied:
it's exposed by a [different PR](https://github.com/h2oai/sparkling-water/pull/2559), which hasn't been merged yet.

michalkurka replied:
thank you - makes sense

michalkurka reviewed (approved):
The concept and implementation look good. 

I suggest to improve scaladoc of the newly added method and explain the difference between the "map" and "object" version.

The weakest point of this PR is concern of backward and forward compatibility - this should be addressed in the future by creating a compatibility testing suite (this suite should check all values are parsed and no warning are generated).

michalkurka commented:
@mn-mikke one more idea - it might be useful to turn the logError/logWarning into errors in tests so that we could easily detect an issue (eg. add a property that switches between an exception and logXY)

mn-mikke commented:
@michalkurka I've added scaladoc and implemented your suggestion about empty metrics. It helped me to identify [some metrics](https://github.com/h2oai/sparkling-water/pull/2565/files#diff-522407b176d7d764fc6eebedd11c18610bb856ccaa2b77d97e468820aa54f434R24) which are not always present in json.
